protocols/gossipsub/src/behaviour.rs

```diff
@@ -18,40 +18,137 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use crate::config::GossipsubConfig;
+use crate::config::{GossipsubConfig, ValidationMode};
+use crate::error::PublishError;
 use crate::handler::GossipsubHandler;
 use crate::mcache::MessageCache;
 use crate::protocol::{
     GossipsubControlAction, GossipsubMessage, GossipsubSubscription, GossipsubSubscriptionAction,
-    MessageId,
+    MessageId, SIGNING_PREFIX,
 };
+use crate::rpc_proto;
 use crate::topic::{Topic, TopicHash};
 use futures::prelude::*;
-use libp2p_core::{Multiaddr, PeerId, connection::ConnectionId};
+use libp2p_core::{
+    connection::ConnectionId, identity::error::SigningError, identity::Keypair, Multiaddr, PeerId,
+};
 use libp2p_swarm::{
-    NetworkBehaviour,
-    NetworkBehaviourAction,
-    NotifyHandler,
-    PollParameters,
-    ProtocolsHandler
+    NetworkBehaviour, NetworkBehaviourAction, NotifyHandler, PollParameters, ProtocolsHandler,
 };
 use log::{debug, error, info, trace, warn};
-use lru::LruCache;
+use lru_time_cache::LruCache;
+use prost::Message;
 use rand;
 use rand::{seq::SliceRandom, thread_rng};
 use std::{
-    collections::hash_map::HashMap,
     collections::HashSet,
     collections::VecDeque,
-    iter,
+    collections::{hash_map::HashMap, BTreeSet},
+    fmt, iter,
     sync::Arc,
     task::{Context, Poll},
 };
 use wasm_timer::{Instant, Interval};
 
 mod tests;
 
+/// Determines if published messages should be signed or not.
+///
+/// Without signing, a number of privacy preserving modes can be selected.
+///
+/// NOTE: The default validation settings are to require signatures. The [`ValidationMode`]
+/// should be updated in the [`GossipsubConfig`] to allow for unsigned messages.
+#[derive(Clone)]
+pub enum MessageAuthenticity {
+    /// Message signing is enabled. The author will be the owner of the key and the sequence number
+    /// will be a random number.
+    Signed(Keypair),
+    /// Message signing is disabled.
+    ///
+    /// The specified `PeerId` will be used as the author of all published messages. The sequence
+    /// number will be randomized.
+    Author(PeerId),
+    /// Message signing is disabled.
+    ///
+    /// A random `PeerId` will be used when publishing each message. The sequence number will be a
+    /// random number.
```

Comment:
```suggestion
    /// A random `PeerId` will be used when publishing each message. The sequence number will be
    /// randomized.
```

Nit pick to align the working with the one above.